### PR TITLE
Create MLOpsBuilder

### LINKED ISSUE #70: Add an MLOpsBuilder to create a configured instance of MLLifeCycleManager
**Is your feature request related to a problem? Please describe.**
The initial setup of `MLLifeCycleManager` for small test scenarios seems problematic in that you are required to set up a backing store before using or it throws an exception. It seems to me there should be a default implementation that requires zero configuration, perhaps an in-memory dictionary based store.

**Describe the solution you'd like**
Add in-memory store implementations of `IMetaDataStore` and `IModelRepository` - these will be useful for testing various scenarios in which the user does not need to persist the data in a permanent storage mechanism. Make these the default implementation instead of throwing an exception via `EnsureStorageProviderConfigured` on every call to `MLLifeCycleManager`

**Describe alternatives you've considered**
Alternatively could use SQLite implementation as default as it does not require any configuration by default, however it does persist on disk which might be undesirable for testing various scenarios.

**Additional context**
N/A

### PR DESCRIPTION
**Pull request by ssa3512**
## Resolves
Fixes #70

### CODE REVIEW

File: src/MLOps.NET/MLLifeCycleManager.cs [hunk 1]
```diff
@@ -8,15 +8,21 @@ namespace MLOps.NET
...
+        {
+            MetaDataStore = metaDataStore ?? throw new ArgumentNullException(nameof(metaDataStore));
+            ModelRepository = modelRepository ?? throw new ArgumentNullException(nameof(modelRepository));
+        }
+
         /// <summary>
         /// Repository for model metadata 
         /// </summary>
-        public IMetaDataStore MetaDataStore { get; set; }
+        public IMetaDataStore MetaDataStore { get; }
```
Comment (aslotte) [inline src/MLOps.NET/MLLifeCycleManager.cs:20]: This is great, lets take it one step further and make them private member variables.

File: src/MLOps.NET.SQLite/MLOpsBuilderExtensions.cs [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+    {
+        /// <summary>
+        /// Enables the usage of SQLite and local storage
+        /// </summary>
+        /// <param name="builder">MLOpsBuilder to add Azure Storage providers to</param>
+        /// <param name="destinationFolder">Destination folder (optional with default value of C:\MLops)</param>
+        /// <returns>Provided MLOpsBuilder for chaining</returns>
+        public static MLOpsBuilder UseSQLite(this MLOpsBuilder builder, string destinationFolder)
+        {
+            builder.UseMetaDataStore(new SQLiteMetaDataStore())
```
Comment (aslotte) [inline src/MLOps.NET.SQLite/MLOpsBuilderExtensions.cs:18]: missing `;`

File: test/MLOps.NET.SQLite.Tests/MLOps.NET.SQLite.Tests.csproj [hunk 3]
```diff
@@ -0,0 +1,7 @@
+<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <TargetFramework>netcoreapp3.1</TargetFramework>
+  </PropertyGroup>
```
Comment (aslotte) [inline test/MLOps.NET.SQLite.Tests/MLOps.NET.SQLite.Tests.csproj:5]: I've started enforcing warnings as errors so you may want to add this in the test project under property group
The CI build doesn't seem to generate warnings for the test project (which is good), but I did have some when building locally.

```
    <!--Do not enforce missing XML warnings on test project -->
    <NoWarn>1591</NoWarn>
```

File: test/MLOps.NET.Tests/MLOps.NET.Tests.csproj [hunk 4]
```diff
@@ -0,0 +1,11 @@
+<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <TargetFramework>netcoreapp3.1</TargetFramework>
```
Comment (aslotte) [inline test/MLOps.NET.Tests/MLOps.NET.Tests.csproj:4]: I've started enforcing warnings as errors so you may want to add this in the test project under property group
The CI build doesn't seem to generate warnings for the test project (which is good), but I did have some when building locally.

```
    <!--Do not enforce missing XML warnings on test project -->
    <NoWarn>1591</NoWarn>
```

File: test/MLOps.NET.Tests/MLOpsBuilderTests.cs [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+namespace MLOps.NET.Tests
+{
+    [TestClass]
+    public class MLOpsBuilderTests
+    {
+        [TestMethod]
+        [ExpectedException(typeof(InvalidOperationException))]
+        public void MLOpsBuilder_ThrowsIfMetaDataStoreConfiguredTwice()
+        {
+            var metaDataStore = new MockMetaDataStore();
```
Comment (aslotte) [inline test/MLOps.NET.Tests/MLOpsBuilderTests.cs:15]: Instead of a Mock, could we use a mocking library instead, e.g. AutoFixture.AutoMoq or alike?

Comment (ssa3512) [reply test/MLOps.NET.Tests/MLOpsBuilderTests.cs]: Added Moq and used that

File: test/MLOps.NET.Tests/MLOps.NET.Tests.csproj (hunk 4 above)
Comment (ssa3512) [reply test/MLOps.NET.Tests/MLOps.NET.Tests.csproj:4]: Added this to the Directory.Build.props in the test folder so it applies to all test projects

File: test/MLOps.NET.SQLite.Tests/MLOps.NET.SQLite.Tests.csproj (hunk 3 above)
Comment (ssa3512) [reply test/MLOps.NET.SQLite.Tests/MLOps.NET.SQLite.Tests.csproj:5]: Added this to the Directory.Build.props in the test folder so it applies to all test projects

File: src/MLOps.NET.SQLite/MLOpsBuilderExtensions.cs (hunk 2 above)
Comment (ssa3512) [reply src/MLOps.NET.SQLite/MLOpsBuilderExtensions.cs]: fixed

File: src/MLOps.NET/MLLifeCycleManager.cs (hunk 1 above)
Comment (ssa3512) [reply src/MLOps.NET/MLLifeCycleManager.cs]: done - I used `lowerCamelCase` for private field names - let me know if you prefer something different (underscore prefix, etc)

File: src/MLOps.NET.Azure/Properties/AssemblyInfo.cs [hunk 6]
```diff
@@ -0,0 +1,3 @@
+﻿using System.Runtime.CompilerServices;
+
+[assembly: InternalsVisibleTo("MLOps.NET.Azure.Tests")]
```
Comment (aslotte) [inline src/MLOps.NET.Azure/Properties/AssemblyInfo.cs:3]: Let's move these to the project file instead

File: src/MLOps.NET.SQLite/Properties/AssemblyInfo.cs [hunk 7]
```diff
@@ -0,0 +1,3 @@
+﻿using System.Runtime.CompilerServices;
+
+[assembly: InternalsVisibleTo("MLOps.NET.SQLite.Tests")]
```
Comment (aslotte) [inline src/MLOps.NET.SQLite/Properties/AssemblyInfo.cs:3]: Same here

File: test/MLOps.NET.Tests/MLOpsBuilderTests.cs [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+            IMLLifeCycleManager lcManager = new MLOpsBuilder()
+                .UseMetaDataStore(metaDataStore)
+                .UseModelRepository(repository)
+                .Build();
+
+            Assert.IsInstanceOfType(lcManager, typeof(MLLifeCycleManager));
+            var metaDataField = typeof(MLLifeCycleManager).GetField("metaDataStore", BindingFlags.Instance | BindingFlags.NonPublic);
+            var repositoryField = typeof(MLLifeCycleManager).GetField("modelRepository", BindingFlags.Instance | BindingFlags.NonPublic);
+
+            Assert.AreSame(metaDataStore, metaDataField.GetValue(lcManager));
```
Comment (aslotte) [inline test/MLOps.NET.Tests/MLOpsBuilderTests.cs:58]: Let's use FluentAssertions for asserts

**Comment by ssa3512** [comment]
This should be ready to go with the exception of figuring out the build - I can build on both windows and WSL2 ubuntu so I am not sure why the CI build is failing.

File: src/MLOps.NET.Azure/Properties/AssemblyInfo.cs (hunk 6 above)
Comment (ssa3512) [reply src/MLOps.NET.Azure/Properties/AssemblyInfo.cs]: Fixed - didn't even know this was a feature with the new csproj format!

File: test/MLOps.NET.Tests/MLOpsBuilderTests.cs (hunk 8 above)
Comment (ssa3512) [reply test/MLOps.NET.Tests/MLOpsBuilderTests.cs]: Done - Never used these before, they seem pretty nice.

Comment (aslotte) [reply test/MLOps.NET.Tests/MLOpsBuilderTests.cs]: I really like FluentAssertions as it makes it a lot more readable, thanks! 👍

**Comment by shanranm** [comment]
@ssa3512 you didn't commit `MLOps.Net.Tests.csproj` after adding FluentAssertion

**Comment by ssa3512** [comment]
@shanranm I don’t think that’s it - the fluent assertions are referenced in the test/Directory.Build.props

**Comment by aslotte** [comment]
@ssa3512 I pulled the branch locally and had a look. I think the build is failing because of the sln file being moved from the src folder to one level up. I think I would like to keep the sln file and the tests + impl projects under the src folder. I may look at doing this tonight if you would like me to

**Comment by ssa3512** [comment]
@aslotte I didn't move the solution on this branch - was experimenting with that on my fork's master branch but it is still under src in this PR. It could be due to the tests being in a `../tests` relative to the sln. I will move them into the src folder and see if that fixes it.

**Comment by ssa3512** [comment]
@aslotte Success!

**Comment by aslotte** [comment]
Fantastic! :)

**Comment by aslotte** [comment]
Thank you for this awesome contribution @ssa3512!

**Pull request merged**